### ISSUE TITLE
HP Spectre x360 internal speakers no sound after bios update and/or board fixing, works on windows

### ISSUE THREAD

**Issue by marian556** [thesofproject/linux#4323]
The Original problem description is at comment 55 of 
https://bugzilla.kernel.org/show_bug.cgi?id=212465#c55
See also attachments in the next comemnts 56,57,58

**Comment by marian556** [comment]
It looks like windows kernel audio drivers are more robust if they can handle bios or hardware board change like this. 
Do you have some low level intel windows audio diagnostics tools that can detect topology on windows and then you can do something similar on Linux?  I have no clue if the lack of speaker audio on Linux is due to bios update or board fixing and messing with board. I have not tried Linux sounds after board fixing and before bios update. I cannot even confirm board or some chip on that was not replaced by another one.  The original Laptop issue that made me to give the Laptop to repair shop was not powering up (dead brick) , so I assume if they changed something it was not audio , it was usb-c power/data chip . I was not told the exact nature of the board problem.

**Comment by plbossart** [comment]
just add this file 
[sof-dyndbg.conf.txt](https://github.com/thesofproject/linux/files/11314963/sof-dyndbg.conf.txt) as /etc/modprobe.d/sof-dyndbg.conf, reboot and attach the dmesg.log.

It's really easy to add an attachment by clicking on "Attach files" in the comment window.

**Comment by marian556** [comment]
sorry small CONFIG_LOG_BUF_SHIFT=18, will try to rebuild kernel
[dmesg.sof.txt](https://github.com/thesofproject/linux/files/11315269/dmesg.sof.txt)

**Comment by plbossart** [comment]
sorry wrong file, use this 
[sof-dyndbg.conf.txt](https://github.com/thesofproject/linux/files/11315398/sof-dyndbg.conf.txt)

**Comment by plbossart** [comment]
you can also add this to /etc/modprobe.d/alsa-base.conf:
blacklist sof-audio-pci-intel-tgl

and then reboot and 'dmesg -C; modprobe sof-audio-pci-intel-tgl'

this will only capture the audio logs.

**Comment by marian556** [comment]
[dmesg.sof.txt](https://github.com/thesofproject/linux/files/11315687/dmesg.sof.txt)

**Comment by marian556** [comment]
I don´t have it
marian@HPSpectre:~/git/linux$ sudo modprobe sof-audio-pci-intel-tgl
modprobe: FATAL: Module sof-audio-pci-intel-tgl not found in directory /lib/modules/6.3.0-rc3+
Can you send me the relevant .config variables for linux kernel compilation , please?

**Comment by marian556** [comment]
[config-6.3.0-rc3+.txt](https://github.com/thesofproject/linux/files/11315768/config-6.3.0-rc3%2B.txt)

**Comment by marian556** [comment]
do you mean
marian@HPSpectre:~/git/linux$ find /lib/modules/6.3.0-rc3+/ | grep snd-sof-pci-intel-tgl
/lib/modules/6.3.0-rc3+/kernel/sound/soc/sof/intel/snd-sof-pci-intel-tgl.ko

**Comment by marian556** [comment]
[dmesg.snd-sof-pci-intel-tgl.txt](https://github.com/thesofproject/linux/files/11315813/dmesg.snd-sof-pci-intel-tgl.txt)

**Comment by marian556** [comment]
Is it this file ?
linux/sound/soc/sof/intel/pci-tgl.c

**Comment by marian556** [comment]
Does not detect tiger lake. Can you tell me the execution path the detection in the kernel it should go through please and I will  hack the code to force it that way.

**Comment by plbossart** [comment]
The results show that the SoundWire machine driver is not selected somehow, even though you have two SoundWire links they are ignored and you end-up with an HDaudio machine driver with HDMI Devices for playback and DMIC for capture.

It's likely a bad BIOS and/or DMI, we have no support for this 

````
Manufacturer:      HP
Product Name:      
Product Version:   
Firmware Version:  F.18
System SKU:        
Board Vendor:      HP
Board Name:        8709
````

**Comment by plbossart** [comment]
@marian556 please try the code in PR #4327

**Comment by marian556** [comment]
Yes it works. Thanks. Good support , although you said above you don´t support :) .  Keep in mind mantra. If it works on windows it should on Linux too. Otherwise the Linux is not better.
Why are the quirks hardcoded? Why don´t you have override runtime feature from the string parameter/argument into module or from an optional configuration file? Default hardcoded compile time heuristics that works for 99.99% users is fine. But you should still be able override runtime. 
If my case is a common scenario that bios entries can change (such as always when updating by following official HP instructions on https://support.hp.com/hk-en/document/ish_4123786-2338478-16)  then it should be hardcoded. If that is something because repair shop replaced a broken chip on my board with a ´compatible´ working one with missing or different IDs, then you should use override runtime facility. Either in a way  ignore bios entries and run hardcoded default detection heuristics as if  DMI_PRODUCT_NAME= "HP Spectre x360 Conv" .  Or pick the parameters  in configuration file or input module string of whatever the module detection heuristics is selecting or mapping to.

**Comment by plbossart** [comment]
We've never seen a case where the DMI information changes. That's a new low. I don't know what your 'repair shop' did but it does look like a botched job. If the kernel needs to discover what the hardware is, we need to have some level of trust in the information reported, and if the system identification changes randomly I am not sure how well this scales.

I'll think of a way to add a kernel parameter to force the detection to use an existing quirk. That will not make things much better because *someone* will have to set that kernel parameter, and that's beyond the ability of most users.

**Comment by marian556** [comment]
I think I have seen on one of your bug tickets (maybe the bug with audio
jack detection problem) mentioning one of you have another HP laptop for
tests?
Can the person try to update the BIOS to the latest HP one to see what
happens? https://support.hp.com/hk-en/document/ish_4123786-2338478-16
I agree with your assessment if the bios information is preserved. Then
really the repair work was botched.
But then again windows handled the botched job well for audio.

**Comment by marian556** [comment]
Also you may want to support hardware enthusiast or hardware hackers who build hybrid systems  in a mix and match way from the compatible components. So some amount of flexibility on a generic kernel via the input module parameters is really desirable.

**Comment by plbossart** [comment]
The premise for SoundWire is that the hardware discovery is based on ACPI.
The initial BIOS vendor already screwed-up the initial bios by listing devices that are NOT present on the hardware, and not listing the ones actually present. That's broken. Windows works around this by ignoring ACPI and having configuration files, it's horrible.
And then the second BIOS updates prevents changes how the system is identified.

We cannot support such cases with a single-build. Windows does not have this requirement, there is a *DIFFERENT* configuration file added for each driver install.

So bottom line is that we are not going to sweat over broken configurations. No one has ever build a hybrid system based on SoundWire because you need hardware know-how to make sure the signal integrity requirements are met and ACPI knowledge to expose the hardware information. SoundWire was never meant to be a plug-and-play bus.
